server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -247,6 +247,7 @@ public InternalEngine(EngineConfig engineConfig) {
             }
             this.lastRefreshedCheckpointListener = new LastRefreshedCheckpointListener(localCheckpointTracker.getProcessedCheckpoint());
             this.internalReaderManager.addListener(lastRefreshedCheckpointListener);
+            this.externalReaderManager.addListener(completionStatsCache);
```

Comment:
I don't like this. These stats live at the `Engine` level, but the reference managers live down here, and we need to register the cache as a refresh listener; if there were any other implementations of `Engine` then they would also need to do the same (except `ReadOnlyEngine`s of course, they don't refresh so don't need to update these stats).

We can't do this in the `Engine` constructor since these reference managers haven't been created when that runs, but I can't see another obvious place in the `Engine` lifecycle to do this. Suggestions welcome.

Reply:
Sure, thanks for the suggestion. Done in c3ae2725aa2a470fccf6b222eae6afbdc557d5c4.